lib/static-server.ts

```diff
@@ -0,0 +1,437 @@
+	'.ogg': 'audio/ogg',
+	'.mp4': 'video/mp4',
+	'.webm': 'video/webm',
+};
+
+export const SERVER_INFO = 'node-static-vendored/1.0';
+
+export class StaticServer {
+	root: string;
+	options: Options;
+	cacheTime: number | null | undefined = 3600;
+	defaultHeaders: Headers = {};
+	defaultExtension = '';
+	constructor(root: string, options?: Options);
+	constructor(options?: Options);
+	constructor(root?: Options | string | null, options?: Options) {
+		if (root && typeof root === 'object') {
+			options = root;
+			root = null;
+		}
+
+		// resolve() doesn't normalize (to lowercase) drive letters on Windows
+		this.root = path.normalize(path.resolve(root || '.'));
+		this.options = options || {};
+
+		this.options.headers ||= {};
+
+		this.options.indexFile ||= 'index.html';
+
+		if ('cacheTime' in this.options) {
+			this.cacheTime = this.options.cacheTime;
+		}
+
+		if ('defaultExtension' in this.options) {
+			this.defaultExtension = `.${this.options.defaultExtension}`;
+		}
+
+		if (this.options.serverInfo !== null) {
+			this.defaultHeaders['server'] = SERVER_INFO;
+		}
+
+		for (const k in this.defaultHeaders) {
+			this.options.headers[k] ||= this.defaultHeaders[k];
+		}
+	}
+
+	async serveDir(
+		pathname: string, req: http.IncomingMessage, res: http.ServerResponse
+	): Promise<Result> {
+		const htmlIndex = path.join(pathname, this.options.indexFile!);
+
+		try {
+			const stat = await fsP.stat(htmlIndex);
+			const status = 200;
+			const headers = {};
+			const originalPathname = decodeURIComponent(new URL(req.url!, 'http://localhost').pathname);
+			if (originalPathname.length && !originalPathname.endsWith('/')) {
+				return this.getResult(301, { 'Location': originalPathname + '/' });
+			} else {
+				return this.respond(status, headers, htmlIndex, stat, req, res);
+			}
+		} catch {
+			return this.getResult(404);
+		}
+	}
+
+	async serveFile(
+		pathname: string, status: number, headers: Headers, req: http.IncomingMessage, res: http.ServerResponse,
+		errorCallback?: ErrorCallback
+	): Promise<Result> {
+		pathname = this.resolve(pathname);
+
+		const stat = await fsP.stat(pathname);
+		const result = await this.respond(status, headers, pathname, stat, req, res);
+		return this.finish(result, req, res, errorCallback);
+	}
+
+	getResult(status: number, headers: Headers = {}, alreadySent = false): Result {
+		return {
+			status,
+			headers,
+			message: http.STATUS_CODES[status],
+			alreadySent,
+		};
+	}
+
+	finish(
+		result: Result, req: http.IncomingMessage, res: http.ServerResponse, errorCallback?: ErrorCallback
+	): Result {
+		if (this.options.serverInfo !== null) {
+			result.headers['server'] ||= SERVER_INFO;
+		}
+
+		// If `alreadySent`, it's been taken care of in `this.stream`.
+		if (!result.alreadySent && !errorCallback?.(result)) {
+			res.writeHead(result.status, result.headers);
+			if (result.status >= 400 || req.method === 'HEAD') {
+				res.write(`${result.status} ${result.message}`);
+			}
```

Comment:
Actually, wait, @Slayer95, I think this is wrong. You're making it so it always sends error messages for `HEAD` requests? Shouldn't it be the other way around?

Reply:
> Actually, wait, @Slayer95, I think this is wrong. You're making it so it always sends error messages for `HEAD` requests? Shouldn't it be the other way around?

You are right. I was sleep-deprived when you asked for suggestions :| Not sure how ``curl --head`` didn't complain.